spec/spec.md

```diff
@@ -607,19 +644,19 @@ sign_and_send_transaction
 {
     “auth_token”: “<auth_token>”,
     “payloads”: [“<transaction>”, ...],
-    “commitment”: “<commitment_level>”,
-    "cluster": "<cluster>",
-    "skip_preflight": <skip_preflight>,
-    "preflight_commitment": "<preflight_commitment>",
+    "options": {
+        “commitment”: “<commitment_level>”,
+        "skip_preflight": <skip_preflight>,
+        "preflight_commitment": "<preflight_commitment>",
+    }
```

Comment:
How important are the `options` (`commitment_level`, `skip_preflight`, `preflight_commitment`)? Are wallets required to honor these? What are the implications of this? For example, do wallets actually want you daps be able to control the preflight (simulation) commitment, or skipping [preflight transaction checks](https://docs.solana.com/developing/clients/jsonrpc-api#sendtransaction) (what exactly does this do?)

Reply:
The preflight options were requested by a wallet developer, to align with the `sendTransaction` options of web3.js and the Solana RPC interface. `commitment` provides the dapp with some control over the finality at which the transaction would be considered successfully sent.

Re: skipping preflight checks, it's in the API description you linked:
> Before submitting, the following preflight checks are performed:
> 
> The transaction signatures are verified
> The transaction is simulated against the bank slot specified by the preflight commitment. On failure an error will be returned. Preflight checks may be disabled if desired. It is recommended to specify the same commitment and preflight commitment to avoid confusing behavior.

Re: including these options, I think I'd pose it as the inverse - our other methods to submit transactions to the RPC nodes (web3.js, RPC API, solana-client) have preflight control options. Is there a good reason to not have them here? I'd not want for a MWA implementation of a dapp to be less capable than a wallet-adapter + web3.js implementation of a dapp.

Reply:
> Is there a good reason to not have them here?

I can think of a few reasons (goodness debatable).

1. Keep initial versions of the API simple.

It's easier to add stuff later than to remove it. The use case for having the dapp provide these is unclear to me. We can always add these options if people really need them. Did the wallet partner provide any guidance about how they expect it to be used?

Also, the longer the wallet waits for confirmation, the longer the dapp will get backgrounded for. Does this matter?

2. The wallet should control the interaction and user security.

This is a bit opinionated, but hear me out.

When a dapp has an RPC connection of its own, it makes sense for it to configure these values. When it's asking the wallet to send things, it makes sense for the wallet to configure these values. Once the wallet gets the signature, from the RPC and returns control to the dapp, the dapp can always wait for a status of `finalized` if it wants to.

Preflight control from the dapp is also not as important if the wallet is simulating the transaction as a distinct step (which it should do!) before having the user sign and submit it. What exactly is the dapp asking to do when it overrides the wallet's defaults? Should wallets be encouraged to let this happen?

3. Anticipate a chain agnostic design.

It seems like multi chain support is going to be important (people are already asking about it). These options are the only very Solana-specific thing in the API. Arguably the network/cluster name is, but the cluster names could be interpreted to mean `solana:<cluster>`.

Building off point 1, it seems better to add this kind of thing later once we have the future API more clearly in mind, unless we know we need it.

Reply:
One other point --

> I'd not want for a MWA implementation of a dapp to be less capable than a wallet-adapter + web3.js implementation of a dapp.

While it's true that WalletAdapter _currently_ has these `SendOptions` and forwards them to browser wallets that accept them, they have been removed from the wallet standard draft for the same reasons as above.

Once a dapp has the broadcast transaction signature from the wallet, it can wait longer than the wallet generally wants to, but not less long. This seems right to me -- the wallet owns the UX when sending, so it should own the configuration.

Reply:
> Did the wallet partner provide any guidance about how they expect it to be used?

I will ask for more context on why they want this

> Preflight control from the dapp is also not as important if the wallet is simulating the transaction as a distinct step

I agree, the case for independent preflight control is shaky. Happy to revisit this with the wallet.

> Also, the longer the wallet waits for confirmation, the longer the dapp will get backgrounded for. Does this matter?

I'd argue that the right thing to do here is to keep the dapp backgrounded until the transaction reaches a desired level of confirmation (in either case, whether the wallet decides the appropriate level or the dapp does). This way, the user gets a consistent "sign and send" experience - they always enter the wallet, approve the transaction, and wait in a wallet UI context until the transaction reaches the desired confirmation, after which they return to the dapp. I say that this UX is desirable because it doesn't leave the "doing the right thing" part to the dapps.

_If_ we still want to allow dapps to have a say in this behavior, keeping just the `commitment` field allows them to configure what level of finality they are comfortable with, and implement their own in-dapp waiting experience by selecting `processed`.

Reply:
Also,

> These options are the only very Solana-specific thing in the API

This is covered in the new "Support for other chains" section at the end. The shape of the `options` object would differ from chain to chain. You could imagine that, in a future rev of the spec that fully specified a second chain, the shape of this object would be extracted and put into a Solana-specific companion spec. The name of the field, `options`, would remain the same, with its shape controlled by the type of association (which encodes the chain in use, e.g. `solana-wallet://`)

Reply:
I discussed with the wallet, and they are OK with dropping the `preflight` parameters from `sign_and_send_transaction`. This will leave only `commitment` in `options`. PR has been updated.